### PR TITLE
Partial evaluation support for hybrid Boolean operations with a dynamic LHS expression

### PR DESCRIPTION
**Pull request by cesarzc**
This change implements partial evaluation support for hybrid Boolean operations with a dynamic LHS expression, including short-circuiting and optimization logic.

### CODE REVIEW

File: compiler/qsc_partial_eval/src/lib.rs [hunk 1]
```diff
@@ -525,6 +525,159 @@ impl<'a> PartialEvaluator<'a> {
...
+            }
+            (_, _) => panic!("invalid Boolean comparison operator {bin_op:?}"),
+        };
+        Ok(result_var)
+    }
+
+    fn eval_logical_bool_bin_op(
+        &mut self,
+        default_result: bool,
+        rhs_eval_condition_var: rir::Variable,
```
Comment (cesarzc) [inline compiler/qsc_partial_eval/src/lib.rs:616]: Consider changing the name to something that conveys that this comes from the LHS of the expression.

File: compiler/qsc_partial_eval/src/lib.rs [hunk 2]
```diff
@@ -525,6 +525,159 @@ impl<'a> PartialEvaluator<'a> {
...
+                // the branch that is evaluated when LHS is false. We cannot use LHS directly as the condition because
+                // the condition to go into the evaluating branch is 'LHS == false', so we need a comparison instruction
+                // whose bound variable will be used as the condition to go into the evaluating branch.
+                let lhs_rir_var = map_eval_var_to_rir_var(lhs_eval_var);
+                let rhs_eval_condition_var_id = self.resource_manager.next_var();
+                let rhs_eval_condition_rir_var = rir::Variable {
+                    variable_id: rhs_eval_condition_var_id,
+                    ty: rir::Ty::Boolean,
+                };
+                let rhs_eval_condition_ins = Instruction::Icmp(
```
Comment (cesarzc) [inline compiler/qsc_partial_eval/src/lib.rs:555]: We could not generate this instruction.

File: compiler/qsc_partial_eval/src/lib.rs (hunk 1 above)
Comment (cesarzc) [reply compiler/qsc_partial_eval/src/lib.rs]: Not needed anymore.

File: compiler/qsc_partial_eval/src/lib.rs (hunk 2 above)
Comment (cesarzc) [reply compiler/qsc_partial_eval/src/lib.rs]: Updated with optimization for ORs.

**Review by swernli — APPROVED** [review]

**Pull request merged**